File: java/src/main/java/falcon/api/event/CustomEvent.java
package falcon.api.event;

import falcon.api.internal.Interop;
import falcon.api.internal.PluginContext;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;

public final class CustomEvent extends Event {
    public CustomEvent(PluginContext context, MemorySegment handle) {
        super(context, handle);
    }

    public String name() {
        return Interop.string(api().eventName(mHandle));
    }

    public String data() {
        return Interop.string(api().eventData(mHandle));
    }

    public void setData(String data) {
        try (Arena arena = Arena.ofConfined()) {
            api().eventSetData(mHandle, Interop.text(arena, data));
        }
    }

    public String source() {
        return eventCause();
    }
}

File: dotnet/Falcon.PluginAPI/Services/CustomEvent.cs
using Falcon.Interop;

namespace Falcon
{
    public sealed unsafe class CustomEvent : Event, IEvent<CustomEvent>
    {
        private CustomEvent(nint handle) : base(handle)
        {
        }

        public static uint EventType
        {
            get
            {
                return FalconConstants.EventCustom;
            }
        }

        public string Name
        {
            get
            {
                return NativeApi.Text(NativeApi.Table->eventName(Raw));
            }
        }

        public string Data
        {
            get
            {
                return NativeApi.Text(NativeApi.Table->eventData(Raw));
            }
            set
            {
                fixed (byte* text = NativeApi.Utf8(value))
                {
                    NativeApi.Table->eventSetData(Raw, text);
                }
            }
        }

        public string Source
        {
            get
            {
                return EventCause();
            }
        }

        public static CustomEvent Create(nint handle)
        {
            return new CustomEvent(handle);
        }
    }
}
